src/import/kv_service.rs

```diff
@@ -115,60 +114,43 @@ impl ImportKv for ImportKVService {
         let import = Arc::clone(&self.importer);
-        let bounded_stream = mpsc::spawn(stream, &self.threads, self.cfg.stream_channel_window);
 
         ctx.spawn(
-            self.threads.spawn(
-                bounded_stream
-                    .into_future()
-                    .map_err(|(e, _)| Error::from(e))
-                    .and_then(move |(chunk, stream)| {
-                        // The first message of the stream specifies the uuid of
-                        // the corresponding engine.
-                        // The engine should be opened before any write.
-                        let head = match chunk {
-                            Some(ref chunk) if chunk.has_head() => chunk.get_head(),
-                            _ => return Err(Error::InvalidChunk),
-                        };
-                        let uuid = Uuid::from_bytes(head.get_uuid())?;
-                        let engine = import.bind_engine(uuid)?;
-                        Ok((engine, stream))
-                    })
-                    .and_then(move |(engine, stream)| {
-                        stream.map_err(Error::from).for_each(move |mut chunk| {
-                            let start = Instant::now_coarse();
-                            if !chunk.has_batch() {
-                                return Err(Error::InvalidChunk);
-                            }
-                            let batch = chunk.take_batch();
-                            let batch_size = engine.write(batch)?;
-                            IMPORT_WRITE_CHUNK_BYTES.observe(batch_size as f64);
-                            IMPORT_WRITE_CHUNK_DURATION.observe(start.elapsed_secs());
-                            Ok(())
-                        })
-                    })
-                    .then(move |res| match res {
-                        Ok(_) => Ok(WriteEngineResponse::new()),
-                        Err(Error::EngineNotFound(v)) => {
-                            let mut resp = WriteEngineResponse::new();
-                            resp.mut_error()
-                                .mut_engine_not_found()
-                                .set_uuid(v.as_bytes().to_vec());
-                            Ok(resp)
-                        }
-                        Err(e) => Err(e),
-                    })
-                    .then(move |res| send_rpc_response!(res, sink, label, timer)),
-            ),
+            self.threads
+                .spawn_fn(move || {
+                    let uuid = Uuid::from_bytes(req.get_uuid())?;
+                    let engine = import.bind_engine(uuid)?;
+                    Ok((engine, req))
+                })
+                .and_then(move |(engine, req)| {
+                    let ts = req.get_commit_ts();
+                    let start = Instant::now_coarse();
+                    let write_size = engine.write(ts, req.get_pairs())?;
+                    IMPORT_WRITE_CHUNK_BYTES.observe(write_size as f64);
```

Comment:
should WRITE_CHUNK now rename to WRITE_PAIR?

Reply:
Let's keep the name to avoid breaking the Grafana.